File: examples/javascript/full-example.mjs
// Hippius S3 — Complete end-to-end example with AWS SDK v3
// Docs: /storage/s3/javascript

import "dotenv/config";
import {
  S3Client,
  CreateBucketCommand,
  PutObjectCommand,
  ListObjectsV2Command,
  GetObjectCommand,
} from "@aws-sdk/client-s3";
import { getSignedUrl } from "@aws-sdk/s3-request-presigner";

// 1. Connect to Hippius S3
const s3 = new S3Client({
  endpoint: "https://s3.hippius.com",
  region: "decentralized",
  credentials: {
    accessKeyId: process.env.HIPPIUS_ACCESS_KEY,
    secretAccessKey: process.env.HIPPIUS_SECRET_KEY,
  },
  forcePathStyle: true,
});

const bucketName = `hippius-demo-js-${Date.now()}`;
const fileKey = "greeting.txt";

// 2. Create bucket
console.log(`Creating bucket: ${bucketName}`);
await s3.send(new CreateBucketCommand({ Bucket: bucketName }));

// 3. Upload string
console.log(`Uploading file: ${fileKey}`);
await s3.send(new PutObjectCommand({
  Bucket: bucketName,
  Key: fileKey,
  Body: "Welcome to Hippius S3 Storage!",
  ContentType: "text/plain",
}));

// 4. List objects
console.log("Listing objects:");
const listRes = await s3.send(new ListObjectsV2Command({ Bucket: bucketName }));
for (const obj of listRes.Contents || []) {
  console.log(`  ${obj.Key} (${obj.Size} bytes)`);
}

// 5. Download file
console.log("Downloading file:");
const getRes = await s3.send(new GetObjectCommand({ Bucket: bucketName, Key: fileKey }));
const content = await getRes.Body.transformToString();
console.log(`  Content: ${content}`);

// 6. Generate presigned URL
console.log("Generating presigned URL:");
const url = await getSignedUrl(
  s3,
  new GetObjectCommand({ Bucket: bucketName, Key: fileKey }),
  { expiresIn: 3600 },
);
console.log(`  URL: ${url}`);

File: examples/python/full_example.py
# Hippius S3 — Complete end-to-end example with boto3
# Docs: /storage/s3/python

import os
from dotenv import load_dotenv
import boto3
from botocore.config import Config

load_dotenv()

# 1. Connect to Hippius S3
s3 = boto3.client(
    "s3",
    endpoint_url="https://s3.hippius.com",
    aws_access_key_id=os.environ["HIPPIUS_ACCESS_KEY"],
    aws_secret_access_key=os.environ["HIPPIUS_SECRET_KEY"],
    region_name="decentralized",
    config=Config(
        signature_version="s3v4",
        s3={"addressing_style": "path"},
    ),
)

import time
bucket_name = f"hippius-demo-py-{int(time.time())}"
file_key = "greeting.txt"

# 2. Create bucket
print(f"Creating bucket: {bucket_name}")
s3.create_bucket(Bucket=bucket_name)

# 3. Upload string
print(f"Uploading file: {file_key}")
s3.put_object(
    Bucket=bucket_name,
    Key=file_key,
    Body=b"Welcome to Hippius S3 Storage!",
)

# 4. List objects
print("Listing objects:")
response = s3.list_objects_v2(Bucket=bucket_name)
for obj in response.get("Contents", []):
    print(f"  {obj['Key']} ({obj['Size']} bytes)")

# 5. Download file
print("Downloading file:")
response = s3.get_object(Bucket=bucket_name, Key=file_key)
content = response["Body"].read().decode("utf-8")
print(f"  Content: {content}")

# 6. Generate presigned URL
print("Generating presigned URL:")
url = s3.generate_presigned_url(
    "get_object",
    Params={"Bucket": bucket_name, "Key": file_key},
    ExpiresIn=3600,
)
print(f"  URL: {url}")
